processor/src/operations/crypto_ops.rs

```diff
@@ -186,6 +190,58 @@ impl Process {
 
         Ok(())
     }
+
+    /// Logs a precompile event by absorbing TAG and COMM into the precompile sponge
+    /// capacity.
+    ///
+    /// Stack transition:
+    /// `[COMM, TAG, PAD, ...] -> [R1, R0, CAP_NEXT, ...]`
+    ///
+    /// Where:
+    /// - The hasher computes: `[CAP_NEXT, R0, R1] = Rpo([CAP_PREV, TAG, COMM])`
+    /// - `CAP_PREV` is the previous sponge capacity provided non-deterministically via helper
+    ///   registers.
+    /// - The VM stack stores each 4-element word in reverse element order, so the top of the stack
+    ///   exposes the elements of `R1` first, followed by the elements of `R0`, then `CAP_NEXT`.
+    pub(super) fn op_log_precompile(&mut self) -> Result<(), ExecutionError> {
+        // Read TAG and COMM from stack, and CAP_PREV from the processor state
+        let comm = self.stack.get_word(0);
+        let tag = self.stack.get_word(4);
+        let cap_prev = self.pc_transcript_state;
+
+        let input_state: HasherState = {
+            let input_state_words = [cap_prev, tag, comm];
+            Word::words_as_elements(&input_state_words).try_into().unwrap()
+        };
+
+        // Perform the RPO permutation, with output state [CAP_NEXT, R0, R1]
+        let (addr, output_state) = self.chiplets.hasher.permute(input_state);
+
+        // Save the hasher address and CAP_PREV in helper registers
+        self.decoder.set_user_op_helpers(
+            Operation::LogPrecompile,
+            &[addr, cap_prev[0], cap_prev[1], cap_prev[2], cap_prev[3]],
+        );
+
+        // Update the processor's precompile sponge capacity with CAP_NEXT
+        let cap_next = Word::from([
+            output_state[STATE_CAP_RANGE.start],
+            output_state[STATE_CAP_RANGE.start + 1],
+            output_state[STATE_CAP_RANGE.start + 2],
+            output_state[STATE_CAP_RANGE.start + 3],
+        ]);
+        self.pc_transcript_state = cap_next;
+
+        // Write the output to stack (top 12 elements) in reverse order: [R1, R0, CAP_NEXT].
+        for i in 0..STATE_WIDTH {
+            self.stack.set(i, output_state[STATE_WIDTH - 1 - i]);
+        }
```

Comment:
This loop makes it hard to understand the ordering of the words on the stack (a la #2059). The commitment and tag are retrieved using `get_word()` - can't we use 3 calls to `set_word()` here? This would make it much clearer what goes where, as well as the ordering.

Reply:
Yes there is - [here](https://github.com/0xMiden/miden-vm/blob/b0254acaefb50ce64ba8502a10f533759bfcfa94/processor/src/processor/mod.rs#L138)

Reply:
`Processor` doesn't seem to be implemented for `Process` though, unless I'm missing something

Reply:
Ah sorry I put the comment on the legacy `Process` implementation, but I really meant to put it [for the `Processor` implementation](https://github.com/0xMiden/miden-vm/blob/476f24b73a664bbe580c5b7a74da51d063e23224/processor/src/processor/operations/crypto_ops.rs#L252-L256). We don't need to do anything about the `Process`'s implementation, indeed!

Reply:
I've updated the processor, it does seem more clear there.